subprojects/gradle/runner/client/src/main/kotlin/com/avito/runner/scheduler/metrics/TestMetricsListenerImpl.kt

```diff
@@ -0,0 +1,97 @@
+package com.avito.runner.scheduler.metrics
+
+import com.avito.logger.LoggerFactory
+import com.avito.logger.create
+import com.avito.runner.service.model.TestCase
+import com.avito.runner.service.model.TestCaseRun
+import com.avito.runner.service.worker.device.Device
+import com.avito.time.TimeProvider
+import java.util.concurrent.ConcurrentHashMap
+
+internal class TestMetricsListenerImpl(
+    private val testMetricsSender: TestMetricsSender,
+    private val timeProvider: TimeProvider,
+    private val loggerFactory: LoggerFactory
+) : TestMetricsListener {
+
+    private val logger = loggerFactory.create<TestMetricsListener>()
+
+    private val onDeviceMarks = ConcurrentHashMap<TestKey, Long>()
+
+    private val startedMarks = ConcurrentHashMap<TestKey, Long>()
+
+    private val finishedMarks = ConcurrentHashMap<TestKey, Long>()
+
+    private var testSuiteStartedTime: Long = 0
+
+    override fun onTestSuiteStarted() {
+        testSuiteStartedTime = timeProvider.nowInMillis()
+    }
+
+    override fun onDevice(
+        device: Device,
+        test: TestCase,
+        targetPackage: String,
+        executionNumber: Int
+    ) {
+        val key = TestKey(test, executionNumber)
+        onDeviceMarks[key] = timeProvider.nowInMillis()
+    }
+
+    override fun started(
+        device: Device,
+        targetPackage: String,
+        test: TestCase,
+        executionNumber: Int
+    ) {
+        val key = TestKey(test, executionNumber)
+        startedMarks[key] = timeProvider.nowInMillis()
+    }
+
+    override fun finished(
+        device: Device,
+        test: TestCase,
+        targetPackage: String,
+        result: TestCaseRun.Result,
+        durationMilliseconds: Long,
+        executionNumber: Int
+    ) {
+        val key = TestKey(test, executionNumber)
+        finishedMarks[key] = timeProvider.nowInMillis()
+    }
+
+    override fun onTestSuiteFinished() {
+        val metricsComputer = TestMetricsComputer(
+            loggerFactory = loggerFactory,
+            testSuiteStartedTime = testSuiteStartedTime,
+            testSuiteEndedTime = timeProvider.nowInMillis(),
+            onDeviceMarks = onDeviceMarks,
+            startedMarks = startedMarks,
+            finishedMarks = finishedMarks
+        )
+
+        with(testMetricsSender) {
+            metricsComputer.computeInitialDelay()
+                ?.let { sendInitialDelay(it) }
+                ?: logger.warn("Not sending initial delay, no data")
```

Comment:
When will it happen? On empty test suite? In such a case can we check it one time but not for all metrics?

Reply:
i see that it leads to coupling (more data leaks from computation to sending)

Reply:
As I understood, we duplicate checks and errors for the same reason - empty test suite.
This idea can be expressed by the code:

```kotlin
class TestSuiteExecutionProfile {
  val totalTime: Long
  val testsExecution: TestsExecution? // <--- one mark to describe the situation
  
  class TestsExecution(
     val initialDelay: Long,
     ...
     val endDelay: Long
  )
}

val profile: TestExecutionProfile = testMetricsComputer.compute()
```

Not a big deal, up to you